### PR TITLE
feat(postgresql): Add json/jsonb type support to PostgresqlDataTypeFa…

### LINKED ISSUE #574: PostgresqlDataTypeFactory support json type
> **Migrated from SourceForge** — [Feature Requests #188](https://sourceforge.net/p/dbunit/feature-requests/188/)
> Originally reported by **hieunv15** on 2015-04-21 13:50:11.692000
> Priority: `5`
> Assigned to **jeffjensen**
> SF status: pending
> Votes: 4

I would like to use PostgreSQL json type with dbunit.
Please support it in next version. I have modified dbunit to support json Please see attachments.
Thanks.

[JsonType.java](https://sourceforge.net/p/dbunit/feature-requests/188/attachment/JsonType.java) (1,293 bytes)
[PostgresqlDataTypeFactory.java](https://sourceforge.net/p/dbunit/feature-requests/188/attachment/PostgresqlDataTypeFactory.java) (4,112 bytes)

---

**jeffjensen** — 2015-04-25 17:54:49.861000
Hi, thanks for adding this.

Please add tests for this feature and each class that prove it works and has no bugs.

For example, manual inspection shows JsonType.setSqlValue will throw NPE when value is null.

### PR DESCRIPTION
**Pull request by jeffjensen**
…ctory

PostgresqlDataTypeFactory had no recognition for PostgreSQL's native json/jsonb columns (reported as sql type OTHER), so they fell through to DefaultDataTypeFactory with no dedicated read/write handling.

Add JsonType, following the same reflection-based PGobject approach as UuidType/InetType/CitextType so dbUnit doesn't need a compile-time dependency on the PostgreSQL driver. One instance handles a single sql type name ("json" or "jsonb") since PostgreSQL has no implicit cast between the two; PostgresqlDataTypeFactory constructs the matching instance per column automatically.

* Unlike the issue reporter's original 2015 patch attachment, which a prior maintainer reply already flagged as NPEing from setSqlValue() on a null value, this implementation explicitly null-checks in both setSqlValue() (binds sql NULL) and typeCast() (returns null).
* Add JsonTypeTest (9 cases, including a Mockito-based regression test for the null setSqlValue() path) and 2 new PostgresqlDataTypeFactoryTest cases.
* Add PostgresqlJsonIT, round-tripping json and jsonb columns (including a null row) through a live PostgreSQL 16 container; jsonb assertions compare parsed document structure since PostgreSQL reformats jsonb text on storage.
* Document JsonType in databases/postgresql.adoc, cross-linking IsActualEqualToExpectedJsonValueComparer for semantic JSON comparison.

Refs: 574

Claude-Session: https://claude.ai/code/session_017rt6YcFdhBCwafLp6n7mcZ

## Summary by Sourcery

Add native PostgreSQL JSON/JSONB column support via a new JsonType and wire it into the PostgreSQL data type factory.

New Features:
- Introduce JsonType to handle PostgreSQL json and jsonb columns as text values without a compile-time dependency on the PostgreSQL driver.
- Have PostgresqlDataTypeFactory recognize json and jsonb sql type names (reported as Types.OTHER) and create matching JsonType instances per column.

Documentation:
- Document JsonType usage and JSON-specific comparison guidance in the PostgreSQL database documentation.

Tests:
- Add unit tests for JsonType covering constructor behavior, type casting, and null handling in setSqlValue().
- Add factory tests ensuring PostgresqlDataTypeFactory returns JsonType for json and jsonb columns.
- Add PostgreSQL integration tests verifying json and jsonb values, including nulls, correctly round-trip through a real database.

## Summary by CodeRabbit

- **New Features**
  - Added support for PostgreSQL `json` and `jsonb` data types.
  - JSON values can be read, inserted, cast, and round-tripped, including `NULL` values.
  - Preserves the specific PostgreSQL type name during database operations.

- **Documentation**
  - Added PostgreSQL JSON/JSONB usage details and guidance for structural JSON comparisons.

### CODE REVIEW

File: src/main/java/org/dbunit/ext/postgresql/JsonType.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        logger.debug("getJson(value={}, connection={}) - start", value,
+                connection);
+
+        Object tempJson = null;
+
+        try
+        {
+            final Class aPGObjectClass = super.loadClass(
+                    "org.postgresql.util.PGobject", connection);
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/ext/postgresql/JsonType.java]: Fixed in 326ef32f - Class/Constructor now use Class<?>/Constructor<?>, and the getMethod/invoke calls use varargs instead of explicit Class[]/Object[] array literals.

File: src/site/asciidoc/databases/postgresql.adoc [hunk 2]
```diff
@@ -9,9 +9,10 @@ equivalent.
 == IDataTypeFactory
 
 link:/dbunit/apidocs/org/dbunit/ext/postgresql/PostgresqlDataTypeFactory.html[PostgresqlDataTypeFactory]
-recognizes `uuid`, `interval`, `inet`, `geometry`, `citext`, `oid`, and
-(via an overridable hook) custom enum types, mapping each to the dedicated
-classes below; everything else delegates to `DefaultDataTypeFactory`.
+recognizes `uuid`, `interval`, `inet`, `geometry`, `citext`, `json`, `jsonb`,
+`oid`, and (via an overridable hook) custom enum types, mapping each to the
+dedicated classes below; everything else delegates to
+`DefaultDataTypeFactory`.
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/site/asciidoc/databases/postgresql.adoc]: Fixed in 326ef32f, adopting the suggested wording.

File: src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
+package org.dbunit.ext.postgresql;
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java:1]: Declined - every existing *IT.java file in this package (PostgresqlUuidIT, PostgresqlUppercaseSchemaIT, PostgresSQLOidIT, SQLHelperDomainPostgreSQLIT, PostgresqlNonPublicSchemaDeleteAllIT) also starts directly with the package declaration, no header. PostgresqlJsonIT matches its actual peer group (this package's other IT tests) rather than JsonType.java/JsonTypeTest.java, which are the ones with the header.

File: src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    @BeforeEach
+    protected void setUp() throws Exception
+    {
+        // Load active postgreSQL profile and connection from Maven pom.xml.
+        _connection = DatabaseEnvironment.getInstance().getConnection();
+        final Statement stat = _connection.getConnection().createStatement();
+        stat.execute("DROP TABLE IF EXISTS " + testTable + ";");
+        stat.execute("CREATE TABLE " + testTable
+                + "(ID INTEGER NOT NULL, DATA json, DATA_B jsonb);");
+        stat.close();
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java]: Try-with-resources applied in 326ef32f for both fixture methods, plus a finally block around the connection close/null in tearDown. Left the SQL-injection (OpenGrep) flag as-is though - it's a hardcoded literal table name (testTable = "json_test") in a test-only DDL string, never influenced by any input, and DDL table/column identifiers aren't bindable via PreparedStatement parameters anyway. Same pattern as the pre-existing PostgresqlUuidIT this file was modeled on.

File: src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                assertThat(col.getSqlTypeName()).as("DATA column sql type name")
+                        .isEqualTo("json");
+            } else if ("DATA_B".equalsIgnoreCase(col.getColumnName()))
+            {
+                assertThat(col.getDataType().getSqlType())
+                        .as("DATA_B column sql type").isEqualTo(Types.OTHER);
+                assertThat(col.getSqlTypeName()).as("DATA_B column sql type name")
+                        .isEqualTo("jsonb");
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/ext/postgresql/PostgresqlJsonIT.java:121]: Fixed in 326ef32f, adopting the suggested dataChecked/dataBChecked tracking.

**Comment by jeffjensen** [comment]
All fixed in 326ef32f unless noted otherwise.

**Applied (nitpick-level, no inline thread to reply on):**
* `JsonTypeTest`: `assertDoesNotThrow` -> AssertJ `assertThatCode(...).as(...).doesNotThrowAnyException()`, matching the rest of the file.
* `PostgresqlJsonIT`: added trailing periods to the 5 `.as()` messages that were missing them.
* `JsonType` constructor: `Objects.requireNonNull(sqlTypeName, ...)` inside the `super()` call instead of a null check after it. Kept the exception type as `NullPointerException` rather than switching to `IllegalArgumentException` (Sourcery's suggestion) - `GenericEnumType`, right next to this class, already does the identical null-guard-on-a-sql-type-name-constructor-arg pattern and throws `NullPointerException`, and `JsonTypeTest` already asserts that type; wanted to stay consistent with the established sibling rather than introduce a second convention.
* `JsonType#getJson()`: collapsed the 5 catch blocks into one `catch (final ReflectiveOperationException e)` (both of you flagged this); removed the now-unused `InvocationTargetException` import.
* `PostgresqlDataTypeFactoryTest`: added `.as()` messages to the 2 new json/jsonb tests.

**Declined:**
* Extracting the PGobject-reflection sequence into a shared helper across `CitextType`/`GenericEnumType`/`InetType`/`UuidType`/`JsonType` - agreed it's real duplication, but 4 of those 5 files predate this PR and are out of its scope; left as a possible future refactor rather than touching untouched, already-shipped classes here.
* `isInstanceOfSatisfying` restructuring in `PostgresqlDataTypeFactoryTest` - kept the existing `isInstanceOf` + cast form instead, since that's what the file's other 7 tests already use; `.as()` messages alone get the consistency win without a shape mismatch against its neighbors.

Individual replies posted on the other inline comments (Class<?>/Constructor<?> typing, the oid/BIGINT doc wording, the metadata-loop assert-nothing gap, try-with-resources) with two declined there too (LGPL header on `PostgresqlJsonIT`, the OpenGrep SQL-injection flag on a hardcoded test DDL string) - reasoning inline on each.

Full unit suite (2070 tests) and the full `postgresql-16` Failsafe IT suite (360 tests) both green after all of the above; `mvnw clean install site` still builds clean with the doc wording change.
